File: mylibrary/src/main/java/com/giosis/library/list/ListActivity.java
package com.giosis.library.list;


import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.giosis.library.R;
import com.giosis.library.bluetooth.BluetoothClass;
import com.giosis.library.main.MainActivity;
import com.giosis.library.util.CommonActivity;
import com.giosis.library.util.DataUtil;
import com.giosis.library.util.ListFragmentFactoryImpl;
import com.giosis.library.util.Preferences;

/**
 * In progress  // Not Upload  //  Today Done
 */
public class ListActivity extends CommonActivity implements OnClickListener,
        ListInProgressFragment2.OnInProgressFragmentListener, ListUploadFailedFragment2.OnFailedCountListener,
        ListTodayDoneFragment2.OnTodayDoneCountListener {

    String TAG = "ListActivity";

    FrameLayout layout_top_back;
    TextView text_top_title;

    LinearLayout layout_list_in_progress;
    TextView text_list_in_progress_count;

    LinearLayout layout_list_upload_failed;
    TextView text_list_upload_failed_count;

    LinearLayout layout_list_today_done;
    TextView text_list_today_done_count;

    ViewPager2 viewpager2_list;
    PagerAdapter2 pagerAdapter2;


    BluetoothClass bluetoothClass;

    /* Fragment numbering */
    public final static int FRAGMENT_PAGE1 = 0;
    public final static int FRAGMENT_PAGE2 = 1;
    public final static int FRAGMENT_PAGE3 = 2;

    @Override
    public void onCountRefresh(int count) {
        text_list_in_progress_count.setText(String.valueOf(count));
    }

    @Override
    public void onFailedCountRefresh(int count) {
        text_list_upload_failed_count.setText(String.valueOf(count));
    }

    @Override
    public void onTodayDoneCountRefresh(int count) {
        text_list_today_done_count.setText(String.valueOf(count));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        getSupportFragmentManager().setFragmentFactory(new ListFragmentFactoryImpl(this));
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list);

        Log.e("Alarm", "ListActivity onCreate   " + Preferences.INSTANCE.getUserId());

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                | WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
                | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);

        layout_top_back = findViewById(R.id.layout_top_back);
        text_top_title = findViewById(R.id.text_top_title);

        text_top_title.setText(getResources().getString(R.string.navi_list));
        layout_top_back.setOnClickListener(this);

        layout_list_in_progress = findViewById(R.id.layout_list_in_progress);
        text_list_in_progress_count = findViewById(R.id.text_list_in_progress_count);
        layout_list_upload_failed = findViewById(R.id.layout_list_upload_failed);
        text_list_upload_failed_count = findViewById(R.id.text_list_upload_failed_count);
        layout_list_today_done = findViewById(R.id.layout_list_today_done);
        text_list_today_done_count = findViewById(R.id.text_list_today_done_count);
        layout_list_in_progress.setOnClickListener(this);
        layout_list_upload_failed.setOnClickListener(this);
        layout_list_today_done.setOnClickListener(this);

        viewpager2_list = findViewById(R.id.viewpager2_list);
        pagerAdapter2 = new PagerAdapter2(this);
        viewpager2_list.setAdapter(pagerAdapter2);


        bluetoothClass = new BluetoothClass(this);

        viewpager2_list.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                super.onPageSelected(position);

                layout_list_in_progress.setSelected(false);
                layout_list_upload_failed.setSelected(false);
                layout_list_today_done.setSelected(false);

                switch (position) {
                    case 0: {
                        layout_list_in_progress.setSelected(true);
                    }
                    break;
                    case 1: {
                        layout_list_upload_failed.setSelected(true);
                    }
                    break;
                    case 2: {
                        layout_list_today_done.setSelected(true);
                    }
                    break;
                    default:
                        break;
                }
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (Preferences.INSTANCE.getUserId().equals("")) {

            Toast.makeText(ListActivity.this, getResources().getString(R.string.msg_qdrive_auto_logout), Toast.LENGTH_SHORT).show();

            try {
                Intent intent;
                if ("SG".equals(Preferences.INSTANCE.getUserNation())) {
                    intent = new Intent(ListActivity.this, Class.forName("com.giosis.util.qdrive.singapore.LoginActivity"));
                } else {
                    intent = new Intent(ListActivity.this, Class.forName("com.giosis.util.qdrive.international.LoginActivity"));
                }
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(intent);
            } catch (Exception ignore) {

            }

        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public void onClick(View v) {

        int id = v.getId();
        if (id == R.id.layout_list_in_progress) {
            viewpager2_list.setCurrentItem(FRAGMENT_PAGE1);

        } else if (id == R.id.layout_list_upload_failed) {
            viewpager2_list.setCurrentItem(FRAGMENT_PAGE2);

        } else if (id == R.id.layout_list_today_done) {
            viewpager2_list.setCurrentItem(FRAGMENT_PAGE3);

        } else if (id == R.id.layout_top_back) {

            DataUtil.inProgressListPosition = 0;
            DataUtil.uploadFailedListPosition = 0;
            Intent intent = new Intent(ListActivity.this, MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
            startActivity(intent);

        }

    }


    public class PagerAdapter2 extends FragmentStateAdapter {

        public PagerAdapter2(@NonNull FragmentActivity fragmentActivity) {
            super(fragmentActivity);
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            switch (position) {
                case 0: {
                    return new ListInProgressFragment2(bluetoothClass);
                }
                case 1:
                    return new ListUploadFailedFragment2();

                case 2:
                    return new ListTodayDoneFragment2(bluetoothClass);

                default:
                    return null;
            }
        }

        @Override
        public int getItemCount() {
            return 3;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        try {
            for (Fragment fragment : getSupportFragmentManager().getFragments()) {
                fragment.onActivityResult(requestCode, resultCode, data);
            }
        } catch (Exception e) {
            Log.e("Exception", TAG + "  onActivityResult Exception : " + e.toString());
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        bluetoothClass.clearBluetoothAdapter();
    }
}

File: app/src/main/java/com/giosis/util/qdrive/singapore/list/ListActivity.kt
package com.giosis.util.qdrive.singapore.list

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.WindowManager
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback
import com.giosis.util.qdrive.singapore.R
import com.giosis.util.qdrive.singapore.bluetooth.BluetoothClass
import com.giosis.util.qdrive.singapore.bluetooth.BluetoothListener
import com.giosis.util.qdrive.singapore.main.MainActivity
import com.giosis.util.qdrive.singapore.util.CommonActivity
import com.giosis.util.qdrive.singapore.util.DataUtil
import com.giosis.util.qdrive.singapore.util.ListFragmentFactoryImpl
import com.giosis.util.qdrive.singapore.util.Preferences
import kotlinx.android.synthetic.main.activity_list.*
import kotlinx.android.synthetic.main.top_title.*

class ListActivity : CommonActivity(), ListInProgressFragment.OnInProgressFragmentListener,
    ListTodayDoneFragment.OnTodayDoneCountListener, ListUploadFailedFragment.OnFailedCountListener {

    var TAG = "ListActivity"

    var bluetoothClass: BluetoothClass? = null

    /* Fragment numbering */
    private val fragmentPage1 = 0
    private val fragmentPage2 = 1
    private val fragmentPage3 = 2


    override fun onCountRefresh(count: Int) {
        text_list_in_progress_count.text = count.toString()
    }

    override fun onFailedCountRefresh(count: Int) {
        text_list_upload_failed_count.text = count.toString()
    }

    override fun onTodayDoneCountRefresh(count: Int) {
        text_list_today_done_count.text = count.toString()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        supportFragmentManager.fragmentFactory = ListFragmentFactoryImpl(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_list)
        Log.e("Alarm", "ListActivity onCreate   " + Preferences.userId)

        window.addFlags(
            WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                    or WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
                    or WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                    or WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
        )

        text_top_title.text = resources.getString(R.string.navi_list)

        layout_in_progress.setOnClickListener {
            viewpager.currentItem = fragmentPage1
        }

        layout_upload_failed.setOnClickListener {
            viewpager.currentItem = fragmentPage2
        }

        layout_today_done.setOnClickListener {
            viewpager.currentItem = fragmentPage3
        }

        layout_top_back.setOnClickListener {
            DataUtil.inProgressListPosition = 0
            DataUtil.uploadFailedListPosition = 0
            val intent = Intent(this@ListActivity, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
            startActivity(intent)
        }

        val pagerAdapter2 = PagerAdapter2(this)
        viewpager.adapter = pagerAdapter2

        bluetoothClass = BluetoothClass(this)

        viewpager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                layout_in_progress.isSelected = false
                layout_upload_failed.isSelected = false
                layout_today_done.isSelected = false
                when (position) {
                    0 -> {
                        layout_in_progress.isSelected = true
                    }
                    1 -> {
                        layout_upload_failed.isSelected = true
                    }
                    2 -> {
                        layout_today_done.isSelected = true
                    }
                    else -> {}
                }
            }
        })

    }

    override fun onResume() {
        super.onResume()

        if (Preferences.userId == "") {
            Toast.makeText(
                this@ListActivity,
                resources.getString(R.string.msg_qdrive_auto_logout),
                Toast.LENGTH_SHORT
            ).show()

            try {
                val intent = Intent(
                    this@ListActivity,
                    Class.forName("com.giosis.util.qdrive.singapore.LoginActivity")
                )
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intent)
            } catch (ignore: Exception) {
            }
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        return if (keyCode == KeyEvent.KEYCODE_BACK) {
            true
        } else {
            super.onKeyDown(keyCode, event)
        }
    }

    inner class PagerAdapter2(fragmentActivity: FragmentActivity) :
        FragmentStateAdapter(fragmentActivity) {
        override fun getItemCount(): Int {
            return 3
        }

        override fun createFragment(position: Int): Fragment {
            return when (position) {
                0 -> {
                    ListInProgressFragment(bluetoothClass as BluetoothListener)
                }
                1 -> ListUploadFailedFragment()
                2 -> ListTodayDoneFragment(bluetoothClass as BluetoothListener)
                else -> throw IllegalArgumentException("unKnown")
            }
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        try {
            for (fragment in supportFragmentManager.fragments) {
                fragment.onActivityResult(requestCode, resultCode, data)
            }
        } catch (e: java.lang.Exception) {
            Log.e("Exception", "$TAG  onActivityResult Exception : $e")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        bluetoothClass!!.clearBluetoothAdapter()
    }
}
